flang/docs/tutorials/addingIntrinsics.md

```diff
@@ -0,0 +1,383 @@
+
+The implementation is done in C++, but there is a hard requirement in not depending on
+the C++ runtime library. That is because it is not wished that Fortran runtime depends
+upon libstdc++ or other C++ runtime libraries. This means that templates and classes
+are OK, `#include <c[library]>` should be OK, but any other includes should be avoided
+unless proven they do not bring C++ runtime in.
+
+The implementation is done in a related `.cpp` file in the same folder. See
+`lib/runtime/character.cpp` for an example.
+
+#### Tests
+To be defined. Ideally, we would want end-to-end tests from Fortran, but it is
+not yet clear where to put them in llvm. Unit tests can also be defined. See
+`unittests/Runtime/character.cpp` for an example.
+
+### Lowering
+- See: `lib/Lower/IntrinsicCall.cpp`
+- Where to submit patch: fir-dev branch of https://github.com/flang-compiler/f18-llvm-project.
+
+#### Implementation
+Lowering of intrinsic is driven by the `IntrinsicLibrary` class. Its member function
+`gen{intrinsic name}` lowers "intrinsic name". These functions have standardized interfaces.
+There is a table in this file that maps an intrinsic name to the function that lowers it.
+To add support for an intrinsic, a new member function containing the lowering implementation
+must be created and mapped into the table.
+
+The lowering implementation can be inlined if it can easily be implemented in a few fir or
+standard mlir dialect operations. Otherwise, the lowering implementation should generate code
+that emits a call to the runtime implementing the intrinsic.
+
+There are three standardized interfaces to choose from for implementation member functions:
+- `mlir::Value genXXX(mlir::Type, llvm::ArrayRef<mlir::Value>)`
+- `fir::ExtendedValue genXXX(mlir::Type, llvm::ArrayRef<fir::ExtendedValue>)`
+- `void genXXX(llvm::ArrayRef<fir::ExtendedValue>)`
+
+The first interface is to be used for elemental numerical and logical intrinsic functions,
+see `genAbs` for an example.
+The second interface is to be used for other intrinsic functions, see `genTrim` for an example.
+The third interface is to be used for intrinsic subroutine, see `genDateAndTime` for an example.
+
+It is worth being familiar with `fir::ExtendedValue`. They described an entity (symbol,
+or evaluated expression) with all its associated dynamic properties (length, extents, lower bounds...).
+A `fir::ExtendedValue` is a union of different SSA value containers that describe a particular category of entities.
+The categories are:
+- `fir::Unboxed` for scalar numericals, logicals, and non-polymorphic derived types without length parameters,
+- `fir::CharBoxValue` for scalar characters,
+- `fir::ArrayBoxValue` for contiguous arrays of numerical, logical, and non-polymorphic derived types without length parameters.
+- `fir::CharArrayBoxValue` contiguous character arrays,
+- `fir::MutableBoxValue` for allocatable and pointers of all types.
+- `fir::BoxValue` for all the rest.
```

Comment:
Some discussion of 'all the rest' might be helpful.  Array sections?  What about `ProcBoxValue`?  Which of these must/might correspond to runtime descriptors?

Reply:
Good point about `ProcBoxValue`. Needs to be there. I will add a few example for  'all the rest'.

> Which of these must/might correspond to runtime descriptors?

I want to stay away from detailing too much the classes here as they might evolve. Pointing to the class headers seems more future proof (the header comments need to improve but that is a separate tasks). Maybe the logic around these value will deserve its own document/illustration at some point.